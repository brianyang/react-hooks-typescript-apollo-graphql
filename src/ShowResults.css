.ant-card.ant-card-bordered {
  margin: 1rem 0;
  border-color: #95989a;
}
.ant-card .ant-card-body {
  display: flex;
  padding: 0;
}
.ant-card-body .row {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  width: 100%;
  align-items: normal;
  justify-content: normal;
}
.first-col.col {
  width: 100%;
  max-width: none;
  text-align: left;
}
.col {
  width: 100%;
  text-align: left;
  border-top: 1px solid #95989a;
  padding: 20px;
  padding-top: 22px;
}
.col > div {
  margin-top: 10px;
}
.col .item-val {
  display: inline;
}

.col .ant-tag {
  border-radius: 0;
  position: absolute;
  top: 0;
  right: 0;
}
@media (min-width: 640px) {
  .first-col.col {
    width: 70%;
    position: relative;
  }
  .col {
    max-width: 15%;
    min-width: 15%;
    width: auto;
    text-align: center;
    border-top: none;
    border-left: 1px solid #95989a;
  }

  .row .ant-form-item {
    min-width: 240px;
    width: auto;
    margin: 0 3rem;
  }
  .row {
    flex-direction: column;
  }

  .col .item-val {
    display: block;
  }
}
