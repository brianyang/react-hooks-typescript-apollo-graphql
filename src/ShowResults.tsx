import React from "react";
import { Card, Tag } from "antd";
import Spinner from "react-spinkit";
import * as T from "./types";
import branding from "./branding";
import "./ShowResults.css";

export const filterStars = (
  filterItems: any,
  items: T.Item[],
  searchStars: any
) => {
  filterItems = items.filter((item: T.Item) => {
    // this bit of code feels like a hack but without implementing a
    // server component there is no elegant way to filter by stars due to github
    // v3 api limitations
    const getOperator = searchStars.split("")[0];
    const hasOperator = isNaN(Number(getOperator));

    if (!hasOperator) return item.stargazers_count === parseInt(searchStars);
    let derivedNumber = searchStars.split("");
    derivedNumber.shift();
    derivedNumber = derivedNumber.join().replace(/,/, "");
    derivedNumber = parseInt(derivedNumber.replace(/,/g, ""));

    if (getOperator === "<") return item.stargazers_count < derivedNumber;
    if (getOperator === ">") return item.stargazers_count > derivedNumber;

    return item;
  });
  return filterItems;
};

export const filterLicense = (
  filterItems: any,
  items: T.Item[],
  searchLicense: any
) => {
  return filterItems.filter(
    (item: T.Item) => item.license && item.license.key === searchLicense
  );
};

export const ShowResults = ({ data }: any) => {
  const { searchStars, searchLicense, searchFork } = data.variables.form;
  if (data.loading) return <Spinner name="circle" fadeIn="none" />;
  let items = data.repo && data.repo.items;

  let filterItems = items;
  if (items) {
    if (data.variables.searchStarsStr) {
      filterItems = filterStars(
        filterItems,
        items,
        data.variables.searchStarsStr
      );
    } else if (searchStars) {
      filterItems = filterStars(filterItems, items, searchStars);
    }

    if (data.variables.searchLicenseStr) {
      filterItems = filterLicense(
        filterItems,
        items,
        data.variables.searchLicenseStr
      );
    } else if (searchLicense) {
      filterItems = filterLicense(filterItems, items, searchLicense);
    }

    if (searchFork !== undefined) {
      filterItems = filterItems.filter((item: T.Item) => {
        if (item.fork === searchFork) return item;
        return true;
      });
    }
  }

  if (filterItems.length === 0)
    return <>There are no results that match your search</>;

  return (
    <>
      {filterItems &&
        filterItems.map((item: T.Item, index: any) => (
          <React.Fragment key={index}>
            <Card>
              <div className="row">
                <span className="first-col col">
                  <h4>
                    <a
                      target="_blank"
                      rel="noopener noreferrer"
                      href={item.html_url}
                    >
                      {item && item.full_name}
                    </a>
                  </h4>
                  <p>{item.description}</p>
                  <Tag color={branding.primary}>
                    {!!item.forks_count && "forked"}
                  </Tag>
                </span>
                <span className="col">
                  <strong>Stars:</strong>
                  <div className="item-val"> {item.stargazers_count}</div>
                </span>
                <span className="col">
                  <strong>License:</strong>
                  <div className="item-val">
                    {item && item.license && item.license.name}
                  </div>
                </span>
              </div>
            </Card>
          </React.Fragment>
        ))}
    </>
  );
};
