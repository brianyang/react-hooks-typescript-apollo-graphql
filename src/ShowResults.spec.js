import React from "react";
import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import renderer from "react-test-renderer";
import { ShowResults, filterStars, filterLicense } from "./ShowResults";

configure({ adapter: new Adapter() });

describe("mount search", () => {
  let wrapper;
  let mockData = {
    variables: {
      form: {
        searchText: "search"
      }
    },
    repo: {
      items: [
        {
          description:
            "A declarative, efficient, and flexible JavaScript library for building user interfaces.",
          fork: false,
          full_name: "facebook/react",
          html_url: "https://github.com/facebook/react",
          id: 10270250
        }
      ]
    }
  };
  beforeEach(() => {
    wrapper = shallow(<ShowResults data={mockData} />);
  });
  it("should render", () => {
    expect(wrapper.exists()).toBeTruthy();
  });
  it("should test filter stars method", () => {
    const filteredStars = filterStars([], mockData.repo.items, "");
    expect(filteredStars).toEqual(mockData.repo.items);
  });
  it("should test filter license method", () => {
    const filteredLicense = filterLicense([], mockData.repo.items, "");
    expect(filteredLicense).toEqual([]);
  });
  it("should render show results markup", () => {
    const tree = renderer.create(<ShowResults data={mockData} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
