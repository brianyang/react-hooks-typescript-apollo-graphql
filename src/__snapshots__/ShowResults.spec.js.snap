// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mount search should render show results markup 1`] = `"There are no results that match your search"`;
